Add `SDFGeometryGenerator` addon.

santi-grau opened the pull request:
New Addon

**Description**

Creates a THREE.BufferGeometry from an SDF.
Added example and documentation

santi-grau committed:
add SDFGeometry ammend

santi-grau committed:
remove comment

santi-grau committed:
Update webgl_geometry_sdf.jpg

santi-grau committed:
pass that check....

Mugen87 commented on examples/jsm/geometries/SDFGeometry.js:
```diff
@@ -0,0 +1,319 @@
+		else if( res == 128 ) [w,h] = [ 2048, 1024 ]
+		else if( res == 256 ) [w,h] = [ 4096, 4096 ]
+		else if( res == 512 ) {
+			if (confirm("Are you sure you can handle 512? It might crash slower devices...")) [w,h] = [ 16384, 8096 ]
```
Confirm dialogs do not belong to geometry generators.

santi-grau replied:
Yeah... generators should not talk that much.

Mugen87 commented on examples/jsm/geometries/SDFGeometry.js:
```diff
@@ -0,0 +1,319 @@
+    }
+`
+
+class SDFtoMesh extends BufferGeometry{
```
If you want to extend the class from `BufferGeometry`, name it `SDFGeometry`.

Maybe you can also think of a different design. Something like `SDFGenerator`. Usage:
```js
const generator = new SDFGeometryGenerator( renderer );

const geometry = generator.generate( res, func, bounds );
```
Since the generator relies on the renderer, materials, shaders and render targets, I would probably vote for this design in order to delimit the class from classic geometry generators which do not have such complex dependencies.

santi-grau replied:
So, if we adopt this design, we would not be extending class, but returning a BufferGeometry on generate()?
Both options sound good to me...

santi-grau replied:
Opted for second option but not sure if makes sense to extend BufferGeometry anymore if this is a generator...
Any nicer name suggestions than SDFGeometryGenerator ? 😅

Mugen87 replied:
When using `SDFGeometryGenerator` I would not extend of `BufferGeometry` anymore. Consider `generate()` like a. factory method than creates geometries. So:
```js
generate(){ let geometry = new BufferGeometry() .... return geometry; }

Mugen87 replied:
> Any nicer name suggestions than SDFGeometryGenerator ? 😅

Yeah, the name is descriptive but quite long. Maybe just `SDFGenerator`?

santi-grau replied:
✅
SDF generator sounds a bit misleading... Let's leave as is, cannot come up w better one 😅

Mugen87 commented on examples/webgl_geometry_sdf.html:
```diff
@@ -0,0 +1,829 @@
+		
+		let renderer = new WebGLRenderer( { alpha : true, preserveDrawingBuffer : true, antialias : false } );
+
+		class AppMain extends HTMLElement{
```
Please do not organize the example like that but implement it like all others. Meaning without classes but simple `init()`, `animate()` `onWindowResize()` functions.

santi-grau replied:
✅

Mugen87 commented on examples/webgl_geometry_sdf.html:
```diff
@@ -0,0 +1,829 @@
+			PlaneGeometry, 
+			Matrix4, 
+			ShaderMaterial 
+		} from 'three';
```
Just do this for simplicity:
```js
import * as THREE from 'three';
```

santi-grau replied:
✅

santi-grau committed:
PR fixes

~ Code scanning checks
~ Update class name to SDFGeometryGenerator
~ Change class design
~ Remove confirm dialog
~ Standardize example structure
~ Remove shader editor

santi-grau committed:
📺🔫

santi-grau commented on examples/webgl_geometry_sdf.html:
```diff
@@ -0,0 +1,829 @@
+				this.mesh.material.wireframe = this.wireframe
+			}
+
+			showEditor(){
```
✅

santi-grau commented on examples/jsm/geometries/SDFGeometry.js:
```diff
@@ -0,0 +1,319 @@
+	constructor( res = 64, distFunc = defaultDist, bounds = 1 ){
+		super()
+
+        this.renderer = new WebGLRenderer( { alpha : true, preserveDrawingBuffer : true, antialias : false } );
```
✅

Mugen87 commented on examples/jsm/geometries/SDFGeometryGenerator.js:
```diff
@@ -0,0 +1,311 @@
+		else if( res == 128 ) [w,h] = [ 2048, 1024 ];
+		else if( res == 256 ) [w,h] = [ 4096, 4096 ];
+		else if( res == 512 ) [w,h] = [ 16384, 8096 ];
+		else return console.warn( 'From MeshCompute : Resolution must be in range 8 < res < 512 and must be ^2');
```
Regarding console logs:

- When the code can't proceed because of wrong input, throw a new error so devs can react via try/catch.
- If the code can proceed but e.g. you have to clamp input values or resize buffers, a warning is appropriate.

Mugen87 replied:
So in this case I would do:
```js
else throw new Error ('THREE. SDFGeometryGenerator: Resolution must be in range 8 < res < 512 and must be ^2' );
```

santi-grau replied:
✅

santi-grau committed:
Update SDFGeometryGenerator.js

- Remove extends Buffergeometry
- Replace logs by error
- Added option for more res if device supports 32k texture
- Remove unused properties

santi-grau committed:
semi missing

gkjohnson commented:
Just a couple thoughts while passing through -

- It looks like this is implementing SDF to geometry functionality with "Surface Nets" which is basically a Marching Cubes alternative. It would be nice to align the API with the current [MarchingCubes](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/objects/MarchingCubes.js) object implementation since in theory surface nets should be a drop in replacement if I'm understanding right. Though the current MarchingCubes class is structured a bit oddly extending from `Mesh` so maybe this can be figured out later.
- It looks like a lot of the code is copy-pasted from [this library file here](https://github.com/mikolalysenko/isosurface/blob/master/lib/surfacenets.js) in another repo. It might make most sense to drop that file (modified to support exports) in the `jsm/libs` directory with the other library files along with a date of copy / version number so it's easier to keep track of.

santi-grau commented:
@gkjohnson thanks for your input.
Regarding moving this to use Marching Cubes, I'm afraid is a bit beyond my skills :( that's why I used Mikola's isosurface.
Are there any docs on how a file at libs should be included? Or just remove from [SDFGeometryGenerator.js](https://github.com/mrdoob/three.js/pull/26837/files/5885c98688a4babbefd882c1329cf97eda305bd6#diff-7284f4f5acf0aa6d2e13c55c16431ccf861ca97206b8ff5c4c497d8001a459c9) and copy at jsm/libs?

gkjohnson commented:
> Regarding moving this to use Marching Cubes

I didn't intend to suggest using MarchingCubes - just that the APIs between the two classes should (some day) be the same since they aim to solve the same problem. I think that can be skipped in this PR.

> Or just remove from [SDFGeometryGenerator.js](https://github.com/mrdoob/three.js/pull/26837/files/5885c98688a4babbefd882c1329cf97eda305bd6#diff-7284f4f5acf0aa6d2e13c55c16431ccf861ca97206b8ff5c4c497d8001a459c9) and copy at jsm/libs?

This one. Ideally the code is formatted the same as the original source so it's easy to diff and update if needed. It will also need to be adjusted to support js exports.

santi-grau committed:
move surface nets to jsm/libs

santi-grau commented:
@gkjohnson moved the surfaceNets script onto jsm/libs

Mugen87 committed:
Update SDFGeometryGenerator.js

Improve code style.

Mugen87 committed:
Update SDFGeometryGenerator.js

Remove `maxTexSize` property.

Mugen87 committed:
Update SDFGeometryGenerator.html

Improve doc page.

Mugen87 committed:
Update webgl_geometry_sdf.html

Mugen87 commented:
@santi-grau I've went over the code to improve the code style and to simplify the example at a few places. The most important thing for me was to avoid custom material shaders in `webgl_geometry_sdf`. So the meshes are now rendered with built-in materials. This will make the example easier to port the example to WebGPU at a later point.

Mugen87 commented:
@santi-grau The PR looks good to me now. The only thing that might be a problem is that the example relies in custom GLSL code. This dependency will make it harder to migrate the demo to `WebGPURenderer` at some point since the GLSL needs to be migrated to TSL to work with both WebGPU and WebGL. 

Would you be okay to add just the Mandelbrot shader and remove the two others? The head and Suzanne shader are so large that rewriting the code to something else will be a pain.

santi-grau commented:
Sure, will update!

santi-grau committed:
Update to example

Remove large glsl examples
Take screenshot

santi-grau commented:
Updated the example removing those two shaders.

Mugen87 committed:
Update SDFGeometryGenerator.js

More clean up.

Mugen87 committed:
Update webgl_geometry_sdf.html

Clean up.

Mugen87 committed:
Update webgl_geometry_sdf.html

Use `dispose()` methods.

Mugen87 committed:
Update webgl_geometry_sdf.html

Avoid using textarea.

Mugen87 committed:
Update webgl_geometry_sdf.html

Improve code style.

Mugen87 committed:
Update SDFGeometryGenerator.js

Fix memory leak.

Mugen87 commented:
@santi-grau While improving the example, I have noticed `SDFGeometryGenerator` makes no use of `dispose()` methods which means the component creates memory leaks. I have refactored the code a bit to fix this issue. You can easily see this when inspecting the `renderer.info` debug object.

Mugen87 commented:
The rendering performance of `computeSDF()` could be noticeably improved by using instanced rendering because the geometry (the plane) is the same for each draw call and only the `tileNum` uniform is different. If `tileNum` and the position of the plane would be an instanced attributes, `computeSDF()` would only require a single draw call, no matter how many tiles are configured.

Would you be willing to give this implementation a try?

Mugen87 committed:
Update SDFGeometryGenerator.js

Simplify usage of `setIndex()`.

Mugen87 commented on examples/jsm/geometries/SDFGeometryGenerator.js:
```diff
@@ -0,0 +1,144 @@
+
+		} );
+		mesh.cells.forEach( p => ids.push( p[ 0 ], p[ 1 ], p[ 2 ] ) );
+		geometry.setAttribute( 'position', new BufferAttribute( new Float32Array( ps ), 4 ) );
```
Why is `position` a buffer attribute that holds four component?

santi-grau replied:
Initially was trying to send AO occlusion from the raymarcher per vertex and was 'hiding' it on this 4th component. 
It has been removed

santi-grau committed:
Remove w from position buffer

santi-grau commented:
> @santi-grau While improving the example, I have noticed `SDFGeometryGenerator` makes no use of `dispose()` methods which means the component creates memory leaks. I have refactored the code a bit to fix this issue. You can easily see the effect when inspecting the `renderer.info` debug object after generating some geometries.

Didn't know we had to manually dispose. Thanks for adding!

santi-grau commented:
> I've realized the rendering performance of `computeSDF()` could be noticeably improved by using instanced rendering because the geometry (the plane) is the same for each draw call and only the `tileNum` uniform is different. If `tileNum` and the position of the plane would be instanced attributes, `computeSDF()` would only require a single draw call, no matter how many tiles are configured.
> 
> Would you be willing to give this implementation a try?

The reason why I'm using 'tile rendering' is to prevent crashes in case the SDF function gets too complex ( e.g. fbm ). This way, no matter what, we are just computing a specific tile size. If you still think that instance rendering will still be safe I can try to implement it.

Mugen87 commented:
> The reason why I'm using 'tile rendering' is to prevent crashes in case the SDF function gets too complex

I see. In this case, I think it's okay to start with the current implemented approach. If users report performance issues, we can revisit this topic (and maybe think about different rendering modes).
